internals/overlord/logstate/manager.go

```diff
@@ -15,26 +15,108 @@
 package logstate
 
 import (
+	"sync"
+
+	"github.com/canonical/pebble/internals/logger"
 	"github.com/canonical/pebble/internals/plan"
 	"github.com/canonical/pebble/internals/servicelog"
 )
 
-type LogManager struct{}
+type LogManager struct {
+	gatherers   map[string]*logGatherer
+	newGatherer func(*plan.LogTarget) (*logGatherer, error)
+	buffers     map[string]*servicelog.RingBuffer
+	plan        *plan.Plan
+	mu          sync.Mutex
+}
 
 func NewLogManager() *LogManager {
-	return &LogManager{}
+	return &LogManager{
+		gatherers:   map[string]*logGatherer{},
+		newGatherer: newLogGatherer,
+		buffers:     map[string]*servicelog.RingBuffer{},
+	}
 }
 
-// PlanChanged is called by the service manager when the plan changes. We stop
-// all running forwarders, and start new forwarders based on the new plan.
+// PlanChanged is called by the service manager when the plan changes. We update the list of gatherers for each forwarder based on the new plan.
 func (m *LogManager) PlanChanged(pl *plan.Plan) {
-	// TODO: implement
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	// Create a map to hold gatherers for the new plan.
+	// Old gatherers will be moved over or deleted.
+	newGatherers := make(map[string]*logGatherer, len(pl.LogTargets))
+
+	for _, target := range pl.LogTargets {
+		gatherer := m.gatherers[target.Name]
+		if gatherer == nil {
+			// Create new gatherer
+			var err error
+			gatherer, err = m.newGatherer(target)
+			if err != nil {
+				logger.Noticef("Internal error: couldn't create gatherer: %v", err)
```

Comment:
I'd recommend, if we are not going to handle this failure here, to remove the error return from newLogGatherer, move the client creation to inside the gatherer.loop, so it can (re)create it as necessary. Not sure why the client itself could error in creation, possibly it doesn't error too?

Reply:
It's an "internal error" because the gatherer creation *should* never fail - all issues *should* be caught when validating the plan.

Client creation could error:
- if the `target.Type` is not valid (impossible to return a client here as we don't know what type of client)
- if the URL is not valid
- if some other configuration option is not valid

Reply:
"Handling" the error here is `continue`-ing on to the next log target. Again, this *should* never happen, so we don't really have a way to handle it other than telling the user what's wrong.